[BI-1238] - Experiments & Observations download UI

mlm483 opened the pull request:
# Description
**Story:** https://breedinginsight.atlassian.net/browse/BI-1238

Clicking the download link on the  Experiments & Observations table now opens a modal allowing the user to select download options. Clicking "Download" in the modal opens the URL `<bi-api url>/v1/programs/{programId}/experiments/{experimentId}/export?{params}` in a new tab. The backend implementation is a separate story, [BI-1465](https://breedinginsight.atlassian.net/browse/BI-1465).

For validation, I chose to show an error notification if no environments are selected (uncheck "All" to test), but this could also be done with form validation on the modal itself.

I used the WAVE browser plugin to evaluate the accessibility.

I updated the radio buttons used by the germplasm list download modal as well to improve accessibility and visual consistency.

# Dependencies
None.

# Testing
You'll need to upload some experimental data to test the UI, message me if you want the files I used.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [ ] I have create/modified unit tests to cover this change
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [x] I have run TAF: https://github.com/Breeding-Insight/taf/actions/runs/4747666470

[BI-1465]: https://breedinginsight.atlassian.net/browse/BI-1465?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mlm483 commented:
I made a dummy controller on the backend for testing, it's in the [bi-api feature/BI-1238 branch](https://github.com/Breeding-Insight/bi-api/tree/feature/BI-1238) if you want to use it.

davedrp commented:
Passed developer testing

dmeidlin commented on src/components/experiments/ExperimentObservationsDownloadButton.vue:
```diff
@@ -0,0 +1,130 @@
+  @Prop()
+  active!: boolean;
+  @Prop()
+  trailDbId!: string;
```
typo

mlm483 replied:
Fixed by https://github.com/Breeding-Insight/bi-web/pull/309/commits/5f5ad541d8769f2b13b0eb598a0fdff09e3e739a.

dmeidlin commented on src/breeding-insight/model/ExperimentEnvironment.ts:
```diff
@@ -0,0 +1,30 @@
+
+import {SelectOption} from "@/breeding-insight/model/SelectOption";
+
+export class ExperimentEnvironment implements SelectOption<string, string> {
```
I suggest renaming this to something more specific, for example `EnvironmentOption`.

mlm483 replied:
Renamed to `EnvironmentOption`, d04009ee.

dmeidlin commented on src/breeding-insight/model/ExperimentObservationsDataset.ts:
```diff
@@ -0,0 +1,30 @@
+
+import {SelectOption} from "@/breeding-insight/model/SelectOption";
+
+export class ExperimentObservationsDataset implements SelectOption<string, string> {
+    static observations = new ExperimentObservationsDataset("Observation Dataset", "observations");
```
I would also consider renaming this class to something like `DatasetOption`.

mlm483 replied:
Renamed to `ExperimentDatasetOption`, `DatasetOption` may not be specific enough; 83997c82.

mlm483 replied:
Done, aaeaed61.

dmeidlin commented on src/breeding-insight/model/FileType.ts:
```diff
@@ -15,7 +15,9 @@
-export class FileType {
+import {SelectOption} from "@/breeding-insight/model/SelectOption";
+
+export class FileType implements SelectOption<string, string> {
```
and for consistency renaming this to `FileTypeOption`

dmeidlin commented on src/components/experiments/ExperimentObservationsDownloadButton.vue:
```diff
@@ -0,0 +1,130 @@
+-->
+
+<template>
+  <section class="experiment-observations-download-button">
```
this will need and id attribute, and as you mentioned, a row value would need to be appended for the id to be unique

mlm483 replied:
I made unique id attributes for this and `GermplasmDownloadButton`,  66044670.

dmeidlin reviewed (changes requested):
Generally looks very good! Using`<fieldset>` for accessibility in the germplasm modal was a good catch!

I recommend factoring out the common parts of `GermplasmDownloadButton` and `ExperimentObservationsDownloadButton` into a new component `DownloadButton`. Also, `SelectModal`(used for germplasm download) and `ExperiementObservationsExportModal` differ only by the selectors and bound variables, so I suggest redefining `SelectModal` to have a slot for the selector content and props for binding to variables, and use the new `SelectModal` in both the germplasm lists table and experiments&observations table.

Here are some styling issues I noticed. 
- after the modal closes, a border appears around the Download link; this border needs to be removed
- the dropdown selector uses a saturated orange value for the background; a light gray already assigned a sass variable should be used instead
- the toggle switch, radio buttons, and checkbox all use the primary purple color; a secondary or tertiary color, like info, should be used instead while keeping the download button purple
- a drop shadow with gradient appears on the different controls depending on the active state; the drop shadow should be removed for consistency with the rest of the UI
- since the environment is a required field, place a red asterisk next to the label and have helper text appear along with the error banner---see the new-ontology-term or new-breeding-method form as a reference

mlm483 commented:
Fixed a UI bug (see screenshot) in GermplasmByList.vue, 95adc80.

![image](https://user-images.githubusercontent.com/128052931/236031134-7aa3b2be-05cb-4202-8ce8-3f05b91ae0c6.png)

mlm483 commented:
RE: [PR review comment](https://github.com/Breeding-Insight/bi-web/pull/309#pullrequestreview-1404897928)

- I factored common code between `ExperimentObservationsDownloadButton` and `GermplasmDownloadButton` into `DownloadButton`. I could go further and utilize `DownloadButton` directly from the list views, but there is quite a bit of markup and a bit of state and logic in `ExperimentObservationsDownloadButton` and `GermplasmDownloadButton` so I think it's best they remain separate components.
- I made `SelectModal` more generic and renamed it `FormModal`, then deleted `ExperiementObservationsExportModal` which was made superfluous.
- I changed the Download button back to an <a> element, so it's appearance is consistent. I created [a new story, BI-1795](https://breedinginsight.atlassian.net/browse/BI-1795), to use buttons that are styled as links instead.
- I made all the inputs more consistent with the existing UI, no more purple, no more drop shadow.
- I marked required inputs with the red asterisk.

dmeidlin reviewed (approved):
Looks good!
